### ISSUE TITLE
Plans for CLI tool?

### ISSUE THREAD

**Issue by mackeyja92** [vfarcic/docker-flow-proxy#27]
Is there currently any plans to implement a cli tool for adding/updating/removing nodes? 

If not, how does it handle duplicates for the same exact path and service name?

**Comment by vfarcic** [comment]
At the moment, there are no plans to create a CLI since HTTP requests serve the same purpose. In a way, cUrl or wget are a CLI for the proxy. A specialized CLI ((instead HTTP requests) can be added if you think that it would fit for your use case better. We'd just need to brainstorm the form of a CLI.

Currently, the proxy supports two types of HTTP requests.

* **reconfigure**: If a service exists, it gets updated. Otherwise, a new entry is added to the proxy configuration. Reconfiguration is idempotent. As such, there will never be a duplicate of the same service name.
* **remove**: A service is removed from the configuration.

At the moment, the proxy does not perform path validation assuming that each service has a unique path(s). It should be relatively easy to add it as a new feature. If a path of a service is the same as a path of a different service, the proxy would return an error code. Please let me know if this would help your use case.

**Comment by mackeyja92** [comment]
A simple CLI tool would be easier than the long curl/wget commands. The simplest solution would just be a python script to generate the HTTP requests and parse the results. 

On the other hand, I could just have my app set itself up on startup. The one concern I have with that is if there is a possibility that when the second instance starts it could cause issues setting the configuration. Or having a lot of reloads on haproxy when nodes come up and down. A way to query and see if a route exists would solve that.

**Comment by vfarcic** [comment]
> A simple CLI tool would be easier than the long curl/wget commands. The simplest solution would just be a python script to generate the HTTP requests and parse the results.

I think you're right. Having a CLI would help. The problem is that I'm a bit tight with time and won't be able to create it before the end of August. Would you like to volunteer? I can provide all the information you'd need.

> The one concern I have with that is if there is a possibility that when the second instance starts it could cause issues setting the configuration.

I just made an update that solves the problem of having multiple instances. When a new instance of the proxy is created, it will have the same configuration as all the others. Also, when a reconfiguration request is sent to one of the instances, it will be propagated to all the others. Please take a look at https://github.com/vfarcic/docker-flow-proxy/blob/master/articles/swarm-mode.md#scaling-the-proxy .

> Or having a lot of reloads on haproxy when nodes come up and down.

There should not be many reloads since there is no need to send "reconfigure" requests more than once per service. You can create new instances of a service without the need to change any configuration. The same holds true in the case of service upgrades. So, the number of reloads is the same as the number of services you are running in production (not instances or upgrades).

**Comment by mackeyja92** [comment]
Sorry, I was a bit unclear. What happens when a second instance of my service starts and tries to add itself to the configuration of the proxy? Would the proxy still reload? Or does it realize there is already an entry and does nothing? 

I will see if i can create the wrapper but not until next week sometime. 

I do like the idea of having a proxy running in global mode sharing info over consul. That is a great improvement and was one of my coworkers concerns about using the proxy.

**Comment by vfarcic** [comment]
> What happens when a second instance of my service starts and tries to add itself to the configuration of the proxy?

Assuming that you are using the proxy in the "swarm" mode, there is no need to do anything with it when the second starts. The proxy uses the new routing mesh introduced in Docker 1.12. It will do reverse proxy and re-send the request to Docker networking which performs load balancing.

> Would the proxy still reload?

You need to reload proxy only when a new service is deployed (not a new instance or a new relesease).

> Or does it realize there is already an entry and does nothing?

Yes it does. The proxy does not send requests to specific IPs but uses Docker networking. All it needs to know is the name of the service. As long as the proxy and the service reside in the same network, there's nothing else to do.


If you'd like to have a chat through Skype or HangOuts, I can explain it in more details.

**Comment by vfarcic** [comment]
Closing ticket due to inactivity. Please reopen it if you feel some additional tasks should be done.
